lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -318,55 +303,60 @@ struct StructuralHasher {
     update(name.getAsOpaquePointer());
   }
 
+  // NOLINTNEXTLINE(misc-no-recursion)
+  void update(Block *block) {
+    for (auto &op : llvm::reverse(*block))
+      update(&op);
+    for (auto type : block->getArgumentTypes())
+      update(type);
+    update(finalizeID(block));
+    update(position);
+    ++position;
+  }
+
+  // NOLINTNEXTLINE(misc-no-recursion)
+  void update(Region *region) {
+    for (auto &block : llvm::reverse(region->getBlocks()))
+      update(&block);
+    update(position);
+    ++position;
+  }
+
   // NOLINTNEXTLINE(misc-no-recursion)
   void update(Operation *op) {
-    if (op->getNumResults())
-      record(op);
-    else if (auto innerSym = dyn_cast<hw::InnerSymbolOpInterface>(op))
-      if (auto attr = innerSym.getInnerSymAttr())
-        if (!attr.empty())
-          record(op);
+    // Hash the regions. We need to make sure an empty region doesn't hash the
+    // same as no region, so we include the number of regions.
+    update(op->getNumRegions());
+    for (auto &region : reverse(op->getRegions()))
+      update(&region);
 
     update(op->getName());
 
-    // Hash the operands.
+    // Record the uses for later hashing.
     for (auto &operand : op->getOpOperands())
       update(operand);
 
-    // Number the block pointers, for use numbering their arguments.
-    for (auto &region : op->getRegions())
-      for (auto &block : region.getBlocks())
-        record(&block);
-
     // This happens after the numbering above, as it uses blockarg numbering
     // for inner symbols.
     update(op, op->getAttrDictionary());
 
-    // Hash the regions. We need to make sure an empty region doesn't hash the
-    // same as no region, so we include the number of regions.
-    update(op->getNumRegions());
-    for (auto &region : op->getRegions()) {
-      update(region.getBlocks().size());
-      for (auto &block : region.getBlocks()) {
-        update(indices.at(&block));
-        for (auto argType : block.getArgumentTypes())
-          update(argType);
-        for (auto &op : block)
-          update(&op);
-      }
-    }
-
     // Record any op results (types).
     for (auto result : op->getResults())
       update(result);
+
+    // Incorporate the hash of uses we have already built.
+    update(finalizeID(op));
+    update(position);
+    ++position;
```

Comment:
I think its fine, but it might make more sense to do a DF numbering (increment this before recursing into the regions, include it right after the op name).

Reply:
I forgot this is doing a backwards walk now